Initial spec-work for single-line-namespace (#4073)
dotnet/csharplang: spec/namespaces.md

## Compilation units

A *compilation_unit* defines the overall structure of a source file. A compilation unit consists of zero or more *using_directive*s followed by zero or more *global_attributes* followed by a *compilation_unit_body*. A *compilation_unit_body* can either be a *single_line_namespace* or zero or more *namespace_member_declaration*s.

```antlr
compilation_unit

>     : extern_alias_directive* using_directive* global_attributes? compilation_unit_body

Reviewer: You'll want to integrate with top-level statements syntax: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/top-level-statements.md#syntax

Reviewer: Never mind, I see that you have a follow-up in OP. I think I like the latter proposal better. Allows to write:
```
// top-level code...

namespace NS;
class C
{
   ...
}
```

    ;

compilation_unit_body
    : namespace_member_declaration*

>     | single_line_namespace

Reviewer: typo: single_line_namespace_declaration

```antlr
compilation_unit

>     : extern_alias_directive* using_directive* global_attributes? (single_line_namespace | (namespace_member_declaration*))

Reviewer 2: Small nit, but I feel like this might read better if pulled out into `namespace_declaration?`, where `namespace_declaration` is an or between `single_line_namespace` and `namespace_member_declaration+`

Author: that's how i started, but that's actually a lot harder to deal with in the spec.  You now need the rules that ensure:

1. a single-line namespace precedes other members.
2. you can only have one single-line namespace
3. you can't have normal namespace follow a single-line namespace
4. a normal namespace can't contain a single-line namespace.

The grammar approach here really enforces all these rules succinctly.

With the above you really just have:

Either:
   1. a single-line namespace (which itself is easily defined)
   2. a normal namespace declaration

That's it :)

Reviewer 2: I don't see how? I've just proposed breaking out what you have here:

```antlr
compilation_unit
    : extern_alias_directive* using_directive* global_attributes? namespace_declaration?

namespace_declaration
	: single_line_namespace
	| namespace_member_declaration+
```

That should be identical to what you have here, but more readable imo.

Author: I see what you mean now.  the problem is that `namespace_declaration` is already a production. We'd have to name this differently.  Perhaps `compilation_unit_body` ?  I would be fine with that name, or something else if you can think of another suitable one.

Reviewer 2: `compilation_unit_body` sounds fine to me.

Author: Done :)

## Namespace declarations

> A *namespace_declaration* has two different forms.  The first consists of the keyword `namespace`, followed by a namespace name and body, optionally followed by a semicolon.  The second consists of the keyword `namespace`, followed by a namespace name, a semicolon and an optional list of directives and declarations.

Author: ```suggestion
A *namespace_declaration* consists of the keyword `namespace`, followed by a namespace name and body, optionally followed by a semicolon.
A *single_line_namespace_declaration* consists of the keyword `namespace`, followed by a namespace name, a semicolon and an optional list of directives and type declarations.
```

```antlr
namespace_declaration
    : 'namespace' qualified_identifier namespace_body ';'?
    ;
    
single_line_namespace_declaration

>     : 'namespace' qualified_identifier ';' extern_alias_directive* using_directive* type_declaration*

Reviewer 2: Consider how top-level statements fit in here as well, since they're in a speclet currently, not in the main spec.

Author: can you link me to that speclet?

The two compilation units contribute to the single global namespace, in this case declaring two classes with the fully qualified names `A` and `B`. Because the two compilation units contribute to the same declaration space, it would have been an error if each contained a declaration of a member with the same name.

> A *single_line_namespace* will contribute members corresponding to the *namespace_declaration* it is semanticall equivalent to.  See ([Namespace Declarations](#namespace-declarations)) for more details.

Author: ```suggestion
A *single_line_namespace* will contribute members corresponding to the *namespace_declaration* it is semantically equivalent to.  See ([Namespace Declarations](#namespace-declarations)) for more details.
```

## Namespace declarations

A *namespace_declaration* consists of the keyword `namespace`, followed by a namespace name and body, optionally followed by a semicolon.
A *single_line_namespace_declaration* consists of the keyword `namespace`, followed by a namespace name, a semicolon and an optional list of directives and type declarations.

```antlr
namespace_declaration
    : 'namespace' qualified_identifier namespace_body ';'?
    ;
    
single_line_namespace_declaration

>     : 'namespace' qualified_identifier ';' extern_alias_directive* using_directive* type_declaration*

Reviewer: What's the point of allowing more using directives after the single-line namespace declaration?

I'd be tempted to do:
```
single_line_namespace_declaration
    : 'namespace' qualified_identifier ';' type_declaration*
    ;
```

Author: > What's the point of allowing more using directives after the single-line namespace declaration?

It's a very core part of this design.  We want the common file case (which has a single namespace, and usings either outside or inside of it) to always be convertible with no change in semantics.

You hit the nail on the head in your followup comment.  We do not want to punish uesrs who put usings inside (which is also a big style that is commonly seen, and i think stylecop pushed people toward).

Reviewer 3: Sometimes I need usings inside to resolve extensions methods.

https://gist.github.com/ufcpp/f30f4876cffea6a42f1b4a810b92b1ff

```
the two namespace declarations above contribute to the same declaration space, in this case declaring two classes with the fully qualified names `N1.N2.A` and `N1.N2.B`. Because the two declarations contribute to the same declaration space, it would have been an error if each contained a declaration of a member with the same name.

A *single_line_namespace_declaration* permits a namespace declaration to be written without the `{ ... }` block.  For example:

```csharp

> namespace Name;

Reviewer 4: Can this appear after usings? If not, there's a subtle difference when you want to convert between the two and the IDE should probably check for that. See this [example](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHABgAIA5AgblIDsIw4oAHCAYzhrwDpqAZkoBvUpUmUcI/JQDCBMRKmqZ0gCyUAsgAoAlMpKqTUgBqUAHpQC8lACZwAZhACuGGA2OmT8gPaMUH4YcPwA6gho8LqW+nSUAPQJfPxmKqYAvumUWSS5zKwc3LzUAsJGalS0XmqyeAr14t6m6jhaeoZNPiYW1naOLu6e2d3+gcGhEVFwMXGJybSpI1K5qrn5LGycPDRKXVKwCK5cMJQWoutMm0U7pRUHMEcnZ2KXzZRAA===).

Author: > Can this appear after usings?

Yes this can.  Will update this example.

Author: actually, going to leave this as is. all we really need to explain is what single-line-namespace is equivalent to.  The externs/usings/attributes that precede it are already explained in the compilation-unit section.

```
the two namespace declarations above contribute to the same declaration space, in this case declaring two classes with the fully qualified names `N1.N2.A` and `N1.N2.B`. Because the two declarations contribute to the same declaration space, it would have been an error if each contained a declaration of a member with the same name.

A *single_line_namespace_declaration* permits a namespace declaration to be written without the `{ ... }` block.  For example:

```csharp
namespace Name;

extern alias A;
using B;

class C
{
}
```
is semantically equivalent to
```csharp
namespace Name
{
    extern alias A;
    using B;

    class C
    {
    }
}
```

> Specifically, *single_line_namespace_declaration* is treated the same as a *namespace_member_declaration* at the same location in the *compilation_unit* with the same *qualified_identifier*.  The *extern_alias_directive*s, *using_directive*s and *type_declaration*s of that *single_line_namespace_declaration* act as if they were declared inside the *namespace_body* of that *namespace_declaration*.

Author: ```suggestion
Specifically, a *single_line_namespace_declaration* is treated the same as a *namespace_member_declaration* at the same location in the *compilation_unit* with the same *qualified_identifier*.  The *extern_alias_directive*s, *using_directive*s and *type_declaration*s of that *single_line_namespace_declaration* act as if they were declared inside the *namespace_body* of that *namespace_declaration*.
```

```
the two namespace declarations above contribute to the same declaration space, in this case declaring two classes with the fully qualified names `N1.N2.A` and `N1.N2.B`. Because the two declarations contribute to the same declaration space, it would have been an error if each contained a declaration of a member with the same name.

A *single_line_namespace_declaration* permits a namespace declaration to be written without the `{ ... }` block.  For example:

```csharp
namespace Name;

extern alias A;
using B;

class C
{
}
```
is semantically equivalent to
```csharp
namespace Name
{
    extern alias A;
    using B;

    class C
    {
    }
}
```

> Specifically, a *single_line_namespace_declaration* is treated the same as a *namespace_member_declaration* at the same location in the *compilation_unit* with the same *qualified_identifier*.  The *extern_alias_directive*s, *using_directive*s and *type_declaration*s of that *single_line_namespace_declaration* act as if they were declared inside the *namespace_body* of that *namespace_declaration*.

Author: ```suggestion
Specifically, a *single_line_namespace_declaration* is treated the same as a *namespace_declaration* at the same location in the *compilation_unit* with the same *qualified_identifier*.  The *extern_alias_directive*s, *using_directive*s and *type_declaration*s of that *single_line_namespace_declaration* act as if they were declared inside the *namespace_body* of that *namespace_declaration*.
```

```
the two namespace declarations above contribute to the same declaration space, in this case declaring two classes with the fully qualified names `N1.N2.A` and `N1.N2.B`. Because the two declarations contribute to the same declaration space, it would have been an error if each contained a declaration of a member with the same name.

A *single_line_namespace_declaration* permits a namespace declaration to be written without the `{ ... }` block.  For example:

```csharp
namespace Name;

extern alias A;
using B;

class C
{
}
```
is semantically equivalent to
```csharp
namespace Name
{
    extern alias A;
    using B;

    class C
    {
    }
}
```

Specifically, a *single_line_namespace_declaration* is treated the same as a *namespace_declaration* at the same location in the *compilation_unit* with the same *qualified_identifier*.  The *extern_alias_directive*s, *using_directive*s and *type_declaration*s of that *single_line_namespace_declaration* act as if they were declared in the same order inside the *namespace_body* of that *namespace_declaration*.

> A source file cannot contain both a *single_line_namespace_declaration* and a *namespace_declaration*.  A source file cannot contain multiple *single_line_namespace_declaration*s

Author: ```suggestion
A source file cannot contain both a *single_line_namespace_declaration* and a *namespace_declaration*.  A source file cannot contain multiple *single_line_namespace_declaration*s.
```

> 

Author: note: the grammar enforces this, so this sentence is technically redundant.  But i thought it was still helpful to call out explicitly.  Not sure the spec's perspective on that.
